File: cocos/3d/misc/create-mesh.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { Mesh, decodeMesh, inflateMesh } from '../assets/mesh';
import { AttributeName, Format, FormatInfos, PrimitiveMode, Attribute } from '../../gfx';
import { Vec3 } from '../../core';
import { IGeometry, IDynamicGeometry, ICreateMeshOptions, ICreateDynamicMeshOptions } from '../../primitive/define';
import { writeBuffer } from './buffer';
import { BufferBlob } from './buffer-blob';

const _defAttrs: Attribute[] = [
    new Attribute(AttributeName.ATTR_POSITION, Format.RGB32F),
    new Attribute(AttributeName.ATTR_NORMAL, Format.RGB32F),
    new Attribute(AttributeName.ATTR_TEX_COORD, Format.RG32F),
    new Attribute(AttributeName.ATTR_TANGENT, Format.RGBA32F),
    new Attribute(AttributeName.ATTR_COLOR, Format.RGBA32F),
];

const v3_1 = new Vec3();
/**
 * @deprecated
 */
export function createMesh (geometry: IGeometry, out?: Mesh, options?: ICreateMeshOptions): Mesh {
    options = options || {};
    // Collect attributes and calculate length of result vertex buffer.
    const attributes: Attribute[] = [];
    let stride = 0;
    const channels: { offset: number; data: number[]; attribute: Attribute; }[] = [];
    let vertCount = 0;

    let attr: Attribute | null;

    const positions = geometry.positions.slice();
    if (positions.length > 0) {
        attr = null;
        if (geometry.attributes) {
            for (const att of geometry.attributes) {
                if (att.name === (AttributeName.ATTR_POSITION as string)) {
                    attr = att;
                    break;
                }
            }
        }

        if (!attr) {
            attr = _defAttrs[0];
        }

        attributes.push(attr);
        const info = FormatInfos[attr.format];
        vertCount = Math.max(vertCount, Math.floor(positions.length / info.count));
        channels.push({ offset: stride, data: positions, attribute: attr });
        stride += info.size;
    }

    if (geometry.normals && geometry.normals.length > 0) {
        attr = null;
        if (geometry.attributes) {
            for (const att of geometry.attributes) {
                if (att.name === (AttributeName.ATTR_NORMAL as string)) {
                    attr = att;
                    break;
                }
            }
        }

        if (!attr) {
            attr = _defAttrs[1];
        }

        const info = FormatInfos[attr.format];
        attributes.push(attr);
        vertCount = Math.max(vertCount, Math.floor(geometry.normals.length / info.count));
        channels.push({ offset: stride, data: geometry.normals, attribute: attr });
        stride += info.size;
    }

    if (geometry.uvs && geometry.uvs.length > 0) {
        attr = null;
        if (geometry.attributes) {
            for (const att of geometry.attributes) {
                if (att.name === (AttributeName.ATTR_TEX_COORD as string)) {
                    attr = att;
                    break;
                }
            }
        }

        if (!attr) {
            attr = _defAttrs[2];
        }

        const info = FormatInfos[attr.format];
        attributes.push(attr);
        vertCount = Math.max(vertCount, Math.floor(geometry.uvs.length / info.count));
        channels.push({ offset: stride, data: geometry.uvs, attribute: attr });
        stride += info.size;
    }

    if (geometry.tangents && geometry.tangents.length > 0) {
        attr = null;
        if (geometry.attributes) {
            for (const att of geometry.attributes) {
                if (att.name === (AttributeName.ATTR_TANGENT as string)) {
                    attr = att;
                    break;
                }
            }
        }

        if (!attr) {
            attr = _defAttrs[3];
        }

        const info = FormatInfos[attr.format];
        attributes.push(attr);
        vertCount = Math.max(vertCount, Math.floor(geometry.tangents.length / info.count));
        channels.push({ offset: stride, data: geometry.tangents, attribute: attr });
        stride += info.size;
    }

    if (geometry.colors && geometry.colors.length > 0) {
        attr = null;
        if (geometry.attributes) {
            for (const att of geometry.attributes) {
                if (att.name === (AttributeName.ATTR_COLOR as string)) {
                    attr = att;
                    break;
                }
            }
        }

        if (!attr) {
            attr = _defAttrs[4];
        }

        const info = FormatInfos[attr.format];
        attributes.push(attr);
        vertCount = Math.max(vertCount, Math.floor(geometry.colors.length / info.count));
        channels.push({ offset: stride, data: geometry.colors, attribute: attr });
        stride += info.size;
    }

    if (geometry.customAttributes) {
        for (let k = 0; k < geometry.customAttributes.length; k++) {
            const ca = geometry.customAttributes[k];
            const info = FormatInfos[ca.attr.format];
            attributes.push(ca.attr);
            vertCount = Math.max(vertCount, Math.floor(ca.values.length / info.count));
            channels.push({ offset: stride, data: ca.values, attribute: ca.attr });
            stride += info.size;
        }
    }

    // Use this to generate final merged buffer.
    const bufferBlob = new BufferBlob();

    // Fill vertex buffer.
    const vertexBuffer = new ArrayBuffer(vertCount * stride);
    const vertexBufferView = new DataView(vertexBuffer);
    for (const channel of channels) {
        writeBuffer(vertexBufferView, channel.data, channel.attribute.format, channel.offset, stride);
    }
    bufferBlob.setNextAlignment(0);
    const vertexBundle: Mesh.IVertexBundle = {
        attributes,
        view: {
            offset: bufferBlob.getLength(),
            length: vertexBuffer.byteLength,
            count: vertCount,
            stride,
        },
    };
    bufferBlob.addBuffer(vertexBuffer);

    // Fill index buffer.
    let indexBuffer: ArrayBuffer | null = null;
    let idxCount = 0;
    const idxStride = 2;
    if (geometry.indices) {
        const { indices } = geometry;
        idxCount = indices.length;
        indexBuffer = new ArrayBuffer(idxStride * idxCount);
        const indexBufferView = new DataView(indexBuffer);
        writeBuffer(indexBufferView, indices, Format.R16UI);
    }

    // Create primitive.
    const primitive: Mesh.ISubMesh = {
        primitiveMode: geometry.primitiveMode || PrimitiveMode.TRIANGLE_LIST,
        vertexBundelIndices: [0],
    };

    if (indexBuffer) {
        bufferBlob.setNextAlignment(idxStride);
        primitive.indexView = {
            offset: bufferBlob.getLength(),
            length: indexBuffer.byteLength,
            count: idxCount,
            stride: idxStride,
        };
        bufferBlob.addBuffer(indexBuffer);
    }

    let minPosition = geometry.minPos;
    if (!minPosition && options.calculateBounds) {
        minPosition = Vec3.set(new Vec3(), Infinity, Infinity, Infinity);
        for (let iVertex = 0; iVertex < vertCount; ++iVertex) {
            Vec3.set(v3_1, positions[iVertex * 3 + 0], positions[iVertex * 3 + 1], positions[iVertex * 3 + 2]);
            Vec3.min(minPosition, minPosition, v3_1);
        }
    }
    let maxPosition = geometry.maxPos;
    if (!maxPosition && options.calculateBounds) {
        maxPosition = Vec3.set(new Vec3(), -Infinity, -Infinity, -Infinity);
        for (let iVertex = 0; iVertex < vertCount; ++iVertex) {
            Vec3.set(v3_1, positions[iVertex * 3 + 0], positions[iVertex * 3 + 1], positions[iVertex * 3 + 2]);
            Vec3.max(maxPosition, maxPosition, v3_1);
        }
    }

    // Create mesh struct.
    const meshStruct: Mesh.IStruct = {
        vertexBundles: [vertexBundle],
        primitives: [primitive],
    };
    if (minPosition) {
        meshStruct.minPosition = new Vec3(minPosition.x, minPosition.y, minPosition.z);
    }
    if (maxPosition) {
        meshStruct.maxPosition = new Vec3(maxPosition.x, maxPosition.y, maxPosition.z);
    }

    // Create mesh.
    if (!out) {
        out = new Mesh();
    }
    out.reset({
        struct: meshStruct,
        data: new Uint8Array(bufferBlob.getCombined()),
    });

    return out;
}

function getPadding (length: number, align: number): number {
    if (align > 0) {
        const remainder = length % align;
        if (remainder !== 0) {
            const padding = align - remainder;
            return padding;
        }
    }

    return 0;
}

function createDynamicMesh (primitiveIndex: number, geometry: IDynamicGeometry, out?: Mesh, options?: ICreateDynamicMeshOptions): Mesh {
    options = options || { maxSubMeshes: 1, maxSubMeshVertices: 1024, maxSubMeshIndices: 1024 };

    const attributes: Attribute[] = [];
    let stream = 0;

    if (geometry.positions.length > 0) {
        attributes.push(new Attribute(AttributeName.ATTR_POSITION, Format.RGB32F, false, stream++, false, 0));
    }

    if (geometry.normals && geometry.normals.length > 0) {
        attributes.push(new Attribute(AttributeName.ATTR_NORMAL, Format.RGB32F, false, stream++, false, 0));
    }

    if (geometry.uvs && geometry.uvs.length > 0) {
        attributes.push(new Attribute(AttributeName.ATTR_TEX_COORD, Format.RG32F, false, stream++, false, 0));
    }

    if (geometry.tangents && geometry.tangents.length > 0) {
        attributes.push(new Attribute(AttributeName.ATTR_TANGENT, Format.RGBA32F, false, stream++, false, 0));
    }

    if (geometry.colors && geometry.colors.length > 0) {
        attributes.push(new Attribute(AttributeName.ATTR_COLOR, Format.RGBA32F, false, stream++, false, 0));
    }

    if (geometry.customAttributes) {
        for (let k = 0; k < geometry.customAttributes.length; k++) {
            const ca = geometry.customAttributes[k];
            const attr = new Attribute();
            attr.copy(ca.attr);
            attr.stream = stream++;
            attributes.push(attr);
        }
    }

    const vertexBundles: Mesh.IVertexBundle[] = [];
    const primitives: Mesh.ISubMesh[] = [];
    let dataSize = 0;

    for (let i = 0; i < options.maxSubMeshes; i++) {
        const primitive: Mesh.ISubMesh = {
            vertexBundelIndices: [],
            primitiveMode: geometry.primitiveMode || PrimitiveMode.TRIANGLE_LIST,
        };

        // add vertex buffers
        for (const attr of attributes) {
            const formatInfo = FormatInfos[attr.format];
            const vertexBufferSize = options.maxSubMeshVertices * formatInfo.size;

            const vertexView: Mesh.IBufferView = {
                offset: dataSize,
                length: vertexBufferSize,
                count: 0,
                stride: formatInfo.size,
            };

            const vertexBundle: Mesh.IVertexBundle = {
                view: vertexView,
                attributes: [attr],
            };

            const vertexBundleIndex = vertexBundles.length;
            primitive.vertexBundelIndices.push(vertexBundleIndex);
            vertexBundles.push(vertexBundle);
            dataSize += vertexBufferSize;
        }

        // add index buffer
        let stride = 0;
        if (geometry.indices16 && geometry.indices16.length > 0) {
            stride = 2;
        } else if (geometry.indices32 && geometry.indices32.length > 0) {
            stride = 4;
        }

        if (stride > 0) {
            dataSize += getPadding(dataSize, stride);
            const indexBufferSize = options.maxSubMeshIndices * stride;

            const indexView: Mesh.IBufferView = {
                offset: dataSize,
                length: indexBufferSize,
                count: 0,
                stride,
            };

            primitive.indexView = indexView;
            dataSize += indexBufferSize;
        }

        primitives.push(primitive);
    }

    const dynamicInfo: Mesh.IDynamicInfo = {
        maxSubMeshes: options.maxSubMeshes,
        maxSubMeshVertices: options.maxSubMeshVertices,
        maxSubMeshIndices: options.maxSubMeshIndices,
    };

    const dynamicStruct: Mesh.IDynamicStruct = {
        info: dynamicInfo,
        bounds: [],
    };
    dynamicStruct.bounds.length = options.maxSubMeshes;

    const meshStruct: Mesh.IStruct = {
        vertexBundles,
        primitives,
        dynamic: dynamicStruct,
    };

    const createInfo: Mesh.ICreateInfo = {
        struct: meshStruct,
        data: new Uint8Array(dataSize),
    };

    if (!out) {
        out = new Mesh();
    }

    out.reset(createInfo);
    out.initialize();
    out.updateSubMesh(primitiveIndex, geometry);

    return out;
}

/**
 * @en mesh utility class, use to create mesh.
 * @zh 网格工具类，用于创建网格。
 */
export class MeshUtils {
    /**
     * @en create a static mesh.
     * @zh 创建一个静态网格。
     * @param geometry @en geometry data use for creating @zh 用于创建的几何数据
     * @param out @en output static mesh @zh 输出的静态网格
     * @param options @en options of creating @zh 创建选项
     * @return @en The created static mesh, which is same as out @zh 新创建的静态网格，同 out 参数
     */
    static createMesh (geometry: IGeometry, out?: Mesh, options?: ICreateMeshOptions): Mesh {
        return createMesh(geometry, out, options);
    }

    /**
     * @en create a dynamic mesh.
     * @zh 创建一个动态网格。
     * @param primitiveIndex @en sub mesh index @zh 子网格索引
     * @param geometry @en geometry data use for creating @zh 用于创建的几何数据
     * @param out @en output dynamic mesh @zh 输出的动态网格
     * @param options @en options of creating @zh 创建选项
     * @return @en The created dynamic mesh, which is same as out @zh 新创建的动态网格，同 out 参数
     */
    static createDynamicMesh (primitiveIndex: number, geometry: IDynamicGeometry, out?: Mesh, options?: ICreateDynamicMeshOptions): Mesh {
        return createDynamicMesh(primitiveIndex, geometry, out, options);
    }

    /**
     * @en decode a mesh.
     *
     * @engineInternal
     */
    static decodeMesh (mesh: Mesh.ICreateInfo): Mesh.ICreateInfo {
        return decodeMesh(mesh);
    }

    /**
     * @en inflate a mesh.
     *
     * @engineInternal
     */
    static inflateMesh (mesh: Mesh.ICreateInfo): Mesh.ICreateInfo {
        return inflateMesh(mesh);
    }
}

export declare namespace createMesh {
    /**
     * @deprecated
     */
    export interface IOptions {
        /**
         * @en calculate mesh's aabb or not
         * @zh 是否计算模型的包围盒。
         */
        calculateBounds?: boolean;
    }
}

File: native/cocos/3d/misc/CreateMesh.cpp
/****************************************************************************
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "3d/misc/CreateMesh.h"
#include <zlib.h>
#include <algorithm>
#include "3d/misc/Buffer.h"
#include "3d/misc/BufferBlob.h"
#include "core/ArrayBuffer.h"
#include "core/DataView.h"
#include "core/assets/RenderingSubMesh.h"
#include "meshopt/meshoptimizer.h"
#include "renderer/gfx-base/GFXDef-common.h"

namespace cc {
namespace {
gfx::AttributeList defAttrs = {
    gfx::Attribute{gfx::ATTR_NAME_POSITION, gfx::Format::RGB32F},
    gfx::Attribute{gfx::ATTR_NAME_NORMAL, gfx::Format::RGB32F},
    gfx::Attribute{gfx::ATTR_NAME_TEX_COORD, gfx::Format::RG32F},
    gfx::Attribute{gfx::ATTR_NAME_TANGENT, gfx::Format::RGBA32F},
    gfx::Attribute{gfx::ATTR_NAME_COLOR, gfx::Format::RGBA32F},
};
} // namespace

Mesh *MeshUtils::createMesh(const IGeometry &geometry, Mesh *out /*= nullptr*/, const ICreateMeshOptions &options /*= {}*/) {
    if (!out) {
        out = ccnew Mesh();
    }

    out->reset(createMeshInfo(geometry, options));
    return out;
}

Mesh::ICreateInfo MeshUtils::createMeshInfo(const IGeometry &geometry, const ICreateMeshOptions &options /* = {}*/) {
    // Collect attributes and calculate length of result vertex buffer.
    gfx::AttributeList attributes;
    uint32_t stride = 0;
    struct Channel {
        uint32_t offset{0};
        ccstd::vector<float> data; // float?
        gfx::Attribute attribute;
    };
    ccstd::vector<Channel> channels;
    uint32_t vertCount = 0;

    const gfx::Attribute *attr = nullptr;

    ccstd::vector<float> positions(geometry.positions);

    if (!positions.empty()) {
        attr = nullptr;
        if (geometry.attributes.has_value()) {
            for (const auto &att : geometry.attributes.value()) {
                if (att.name == gfx::ATTR_NAME_POSITION) {
                    attr = &att;
                    break;
                }
            }
        }

        if (attr == nullptr) {
            attr = &defAttrs[0];
        }

        attributes.emplace_back(*attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
        vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(positions.size() / info.count)));
        channels.emplace_back(Channel{stride, positions, *attr});
        stride += info.size;
    }

    if (geometry.normals.has_value() && !geometry.normals.value().empty()) {
        attr = nullptr;
        if (geometry.attributes.has_value()) {
            for (const auto &att : geometry.attributes.value()) {
                if (att.name == gfx::ATTR_NAME_NORMAL) {
                    attr = &att;
                    break;
                }
            }
        }

        if (attr == nullptr) {
            attr = &defAttrs[1];
        }

        attributes.emplace_back(*attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
        vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(geometry.normals->size() / info.count)));
        channels.emplace_back(Channel{stride, geometry.normals.value(), *attr});
        stride += info.size;
    }

    if (geometry.uvs.has_value() && !geometry.uvs.value().empty()) {
        attr = nullptr;
        if (geometry.attributes.has_value()) {
            for (const auto &att : geometry.attributes.value()) {
                if (att.name == gfx::ATTR_NAME_TEX_COORD) {
                    attr = &att;
                    break;
                }
            }
        }

        if (attr == nullptr) {
            attr = &defAttrs[2];
        }

        attributes.emplace_back(*attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
        vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(geometry.uvs->size() / info.count)));
        channels.emplace_back(Channel{stride, geometry.uvs.value(), *attr});
        stride += info.size;
    }

    if (geometry.tangents.has_value() && !geometry.tangents.value().empty()) {
        attr = nullptr;
        if (geometry.attributes.has_value()) {
            for (const auto &att : geometry.attributes.value()) {
                if (att.name == gfx::ATTR_NAME_TANGENT) {
                    attr = &att;
                    break;
                }
            }
        }

        if (attr == nullptr) {
            attr = &defAttrs[3];
        }

        attributes.emplace_back(*attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
        vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(geometry.tangents->size() / info.count)));
        channels.emplace_back(Channel{stride, geometry.tangents.value(), *attr});
        stride += info.size;
    }

    if (geometry.colors.has_value() && !geometry.colors.value().empty()) {
        attr = nullptr;
        if (geometry.attributes.has_value()) {
            for (const auto &att : geometry.attributes.value()) {
                if (att.name == gfx::ATTR_NAME_COLOR) {
                    attr = &att;
                    break;
                }
            }
        }

        if (attr == nullptr) {
            attr = &defAttrs[4];
        }

        attributes.emplace_back(*attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
        vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(geometry.colors->size() / info.count)));
        channels.emplace_back(Channel{stride, geometry.colors.value(), *attr});
        stride += info.size;
    }

    if (geometry.customAttributes.has_value()) {
        for (const auto &ca : geometry.customAttributes.value()) {
            const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr->format)];
            attributes.emplace_back(ca.attr);
            vertCount = std::max(vertCount, static_cast<uint32_t>(std::floor(ca.values.size() / info.count)));
            channels.emplace_back(Channel{stride, ca.values, ca.attr});
            stride += info.size;
        }
    }

    // Use this to generate final merged buffer.
    BufferBlob bufferBlob;

    // Fill vertex buffer.
    auto *vertexBuffer = ccnew ArrayBuffer(vertCount * stride);
    DataView vertexBufferView(vertexBuffer);
    for (const auto &channel : channels) {
        writeBuffer(vertexBufferView, channel.data, channel.attribute.format, channel.offset, stride);
    }
    bufferBlob.setNextAlignment(0);
    Mesh::IVertexBundle vertexBundle;
    Mesh::IBufferView buffferView;

    buffferView.offset = bufferBlob.getLength();
    buffferView.length = static_cast<uint32_t>(vertexBuffer->byteLength());
    buffferView.count = vertCount;
    buffferView.stride = stride;
    vertexBundle.attributes = attributes;
    vertexBundle.view = buffferView;

    bufferBlob.addBuffer(vertexBuffer);

    // Fill index buffer.
    ArrayBuffer::Ptr indexBuffer;
    uint32_t idxCount = 0;
    const uint32_t idxStride = 2;
    if (geometry.indices.has_value()) {
        const ccstd::vector<uint32_t> &indices = geometry.indices.value();
        idxCount = static_cast<uint32_t>(indices.size());
        indexBuffer = ccnew ArrayBuffer(idxStride * idxCount);
        DataView indexBufferView(indexBuffer);
        writeBuffer(indexBufferView, indices, gfx::Format::R16UI);
    }

    // Create primitive.
    Mesh::ISubMesh primitive;
    primitive.vertexBundelIndices = {0};
    primitive.primitiveMode = geometry.primitiveMode.has_value() ? geometry.primitiveMode.value() : gfx::PrimitiveMode::TRIANGLE_LIST;

    if (indexBuffer) {
        bufferBlob.setNextAlignment(idxStride);
        Mesh::IBufferView bufferView;
        bufferView.offset = bufferBlob.getLength();
        bufferView.length = indexBuffer->byteLength();
        bufferView.count = idxCount;
        bufferView.stride = idxStride;
        primitive.indexView = bufferView;
        bufferBlob.addBuffer(indexBuffer);
    }

    ccstd::optional<Vec3> minPosition = geometry.minPos;
    if (!minPosition.has_value() && options.calculateBounds.has_value() && options.calculateBounds.value()) {
        minPosition = Vec3(std::numeric_limits<float>::infinity(), std::numeric_limits<float>::infinity(), std::numeric_limits<float>::infinity());
        for (uint32_t iVertex = 0; iVertex < vertCount; ++iVertex) {
            Vec3::min(minPosition.value(), Vec3(positions[iVertex * 3 + 0], positions[iVertex * 3 + 1], positions[iVertex * 3 + 2]), &minPosition.value());
        }
    }

    ccstd::optional<Vec3> maxPosition = geometry.maxPos;
    if (!maxPosition.has_value() && options.calculateBounds.has_value() && options.calculateBounds.value()) {
        maxPosition = Vec3(-std::numeric_limits<float>::infinity(), -std::numeric_limits<float>::infinity(), -std::numeric_limits<float>::infinity());
        for (uint32_t iVertex = 0; iVertex < vertCount; ++iVertex) {
            Vec3::max(maxPosition.value(), Vec3(positions[iVertex * 3 + 0], positions[iVertex * 3 + 1], positions[iVertex * 3 + 2]), &maxPosition.value());
        }
    }

    // Create mesh struct
    Mesh::IStruct meshStruct;
    meshStruct.vertexBundles = {vertexBundle};
    meshStruct.primitives = {primitive};

    if (minPosition.has_value()) {
        meshStruct.minPosition = minPosition.value();
    }
    if (maxPosition.has_value()) {
        meshStruct.maxPosition = maxPosition.value();
    }

    Mesh::ICreateInfo createInfo;
    createInfo.structInfo = std::move(meshStruct);
    createInfo.data = Uint8Array(bufferBlob.getCombined());
    return createInfo;
}

static inline uint32_t getPadding(uint32_t length, uint32_t align) {
    if (align > 0U) {
        const uint32_t remainder = length % align;
        if (remainder != 0U) {
            const uint32_t padding = align - remainder;
            return padding;
        }
    }

    return 0U;
}

Mesh *MeshUtils::createDynamicMesh(index_t primitiveIndex, const IDynamicGeometry &geometry, Mesh *out /*= nullptr*/, const ICreateDynamicMeshOptions &options /*= {}*/) {
    if (!out) {
        out = ccnew Mesh();
    }

    out->reset(MeshUtils::createDynamicMeshInfo(geometry, options));
    out->initialize();
    out->updateSubMesh(primitiveIndex, geometry);

    return out;
}

Mesh::ICreateInfo MeshUtils::createDynamicMeshInfo(const IDynamicGeometry &geometry, const ICreateDynamicMeshOptions &options /* = {}*/) {
    gfx::AttributeList attributes;
    uint32_t stream = 0U;

    if (!geometry.positions.empty()) {
        attributes.push_back({gfx::ATTR_NAME_POSITION, gfx::Format::RGB32F, false, stream++, false, 0U});
    }

    if (geometry.normals.has_value() && !geometry.normals.value().empty()) {
        attributes.push_back({gfx::ATTR_NAME_NORMAL, gfx::Format::RGB32F, false, stream++, false, 0U});
    }

    if (geometry.uvs.has_value() && !geometry.uvs.value().empty()) {
        attributes.push_back({gfx::ATTR_NAME_TEX_COORD, gfx::Format::RG32F, false, stream++, false, 0U});
    }

    if (geometry.tangents.has_value() && !geometry.tangents.value().empty()) {
        attributes.push_back({gfx::ATTR_NAME_TANGENT, gfx::Format::RGBA32F, false, stream++, false, 0U});
    }

    if (geometry.colors.has_value() && !geometry.colors.value().empty()) {
        attributes.push_back({gfx::ATTR_NAME_COLOR, gfx::Format::RGBA32F, false, stream++, false, 0U});
    }

    if (geometry.customAttributes.has_value()) {
        for (const auto &ca : geometry.customAttributes.value()) {
            auto attr = ca.attr;
            attr.stream = stream++;
            attributes.emplace_back(attr);
        }
    }

    ccstd::vector<Mesh::IVertexBundle> vertexBundles;
    ccstd::vector<Mesh::ISubMesh> primitives;
    uint32_t dataSize = 0U;

    for (auto i = 0U; i < options.maxSubMeshes; i++) {
        Mesh::ISubMesh primitive;
        primitive.primitiveMode = geometry.primitiveMode.has_value() ? geometry.primitiveMode.value() : gfx::PrimitiveMode::TRIANGLE_LIST;

        // add vertex buffers
        for (const auto &attr : attributes) {
            const auto &formatInfo = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attr.format)];
            uint32_t vertexBufferSize = options.maxSubMeshVertices * formatInfo.size;

            Mesh::IBufferView vertexView = {
                dataSize,
                vertexBufferSize,
                0U,
                formatInfo.size};

            Mesh::IVertexBundle vertexBundle = {
                0U,
                vertexView,
                {attr}};

            const auto vertexBundleIndex = static_cast<uint32_t>(vertexBundles.size());
            primitive.vertexBundelIndices.emplace_back(vertexBundleIndex);
            vertexBundles.emplace_back(vertexBundle);
            dataSize += vertexBufferSize;
        }

        // add index buffer
        uint32_t stride = 0U;
        if (geometry.indices16.has_value() && !geometry.indices16.value().empty()) {
            stride = sizeof(uint16_t);
        } else if (geometry.indices32.has_value() && !geometry.indices32.value().empty()) {
            stride = sizeof(uint32_t);
        }

        if (stride > 0U) {
            dataSize += getPadding(dataSize, stride);
            uint32_t indexBufferSize = options.maxSubMeshIndices * stride;

            Mesh::IBufferView indexView = {
                dataSize,
                indexBufferSize,
                0U,
                stride};

            primitive.indexView = indexView;
            dataSize += indexBufferSize;
        }

        primitives.emplace_back(primitive);
    }

    Mesh::IDynamicInfo dynamicInfo = {options.maxSubMeshes,
                                      options.maxSubMeshVertices,
                                      options.maxSubMeshIndices};

    Mesh::IDynamicStruct dynamicStruct;
    dynamicStruct.info = dynamicInfo;
    dynamicStruct.bounds.resize(options.maxSubMeshes);
    for (auto &bound : dynamicStruct.bounds) {
        bound.setValid(false);
    }

    Mesh::IStruct meshStruct;
    meshStruct.vertexBundles = vertexBundles;
    meshStruct.primitives = primitives;
    meshStruct.dynamic = std::move(dynamicStruct);

    Mesh::ICreateInfo createInfo;
    createInfo.structInfo = std::move(meshStruct);
    createInfo.data = Uint8Array(dataSize);
    return createInfo;
}

void MeshUtils::inflateMesh(const Mesh::IStruct &structInfo, Uint8Array &data) {
    uLongf uncompressedSize = 0U;
    for (const auto &prim : structInfo.primitives) {
        if (prim.indexView.has_value()) {
            uncompressedSize += prim.indexView->length + prim.indexView->stride;
        }
        if (prim.cluster.has_value()) {
            uncompressedSize += prim.cluster->vertexView.length + prim.cluster->vertexView.stride;
            uncompressedSize += prim.cluster->triangleView.length + prim.cluster->triangleView.stride;
            uncompressedSize += prim.cluster->clusterView.length + prim.cluster->clusterView.stride;
            uncompressedSize += prim.cluster->coneView.length + prim.cluster->coneView.stride;
        }
    }
    for (const auto &vb : structInfo.vertexBundles) {
        uncompressedSize += vb.view.length + vb.view.stride;
    }
    auto uncompressedData = Uint8Array(static_cast<uint32_t>(uncompressedSize));
    auto res = uncompress(uncompressedData.buffer()->getData(), &uncompressedSize, data.buffer()->getData(), data.byteLength());
    data = Uint8Array(uncompressedData.buffer(), 0, static_cast<uint32_t>(uncompressedSize));
}

void MeshUtils::decodeMesh(Mesh::IStruct &structInfo, Uint8Array &data) {
    BufferBlob bufferBlob;

    for (auto &bundle : structInfo.vertexBundles) {
        auto &view = bundle.view;
        auto bound = view.count * view.stride;
        auto *buffer = ccnew ArrayBuffer(bound);
        auto vertex = Uint8Array(data.buffer(), view.offset, view.length);
        int res = meshopt_decodeVertexBuffer(buffer->getData(), view.count, view.stride, vertex.buffer()->getData() + vertex.byteOffset(), view.length);
        if (res < 0) {
            assert(false && "failed to decode vertex buffer");
        }

        bufferBlob.setNextAlignment(view.stride);
        Mesh::IVertexBundle vertexBundle;
        Mesh::IBufferView buffferView;
        buffferView.offset = bufferBlob.getLength();
        buffferView.length = bound;
        buffferView.count = view.count;
        buffferView.stride = view.stride;
        bufferBlob.addBuffer(buffer);

        bundle.view = buffferView;
    }

    for (auto &primitive : structInfo.primitives) {
        if (!primitive.indexView.has_value()) {
            continue;
        }

        auto view = *primitive.indexView;
        auto bound = view.count * view.stride;
        auto *buffer = ccnew ArrayBuffer(bound);
        auto index = DataView(data.buffer(), view.offset, view.length);
        int res = meshopt_decodeIndexBuffer(buffer->getData(), view.count, view.stride, index.buffer()->getData() + index.byteOffset(), view.length);
        if (res < 0) {
            assert(false && "failed to decode index buffer");
        }

        bufferBlob.setNextAlignment(view.stride);
        Mesh::IBufferView buffferView;
        buffferView.offset = bufferBlob.getLength();
        buffferView.length = bound;
        buffferView.count = view.count;
        buffferView.stride = view.stride;
        bufferBlob.addBuffer(buffer);

        primitive.indexView = buffferView;
    }

    data = Uint8Array(bufferBlob.getCombined());
}

} // namespace cc
